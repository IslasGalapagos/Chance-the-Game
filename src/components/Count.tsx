/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import * as React from 'react';
import {StyledWrapper, StyledNum} from './Count.styles';

const enum GameState {
  Draw = 'draw',
  User = 'user_lead',
  Random = 'random_lead'
}

type Score = [number, number];

interface CountProps {
  score: Score;
}

class Count extends React.PureComponent<CountProps, {}> {
  getGameState = (score: Score): GameState => {
    const diff = score[0] - score[1];

    switch (diff === 0 ? 0 : diff / Math.abs(diff)) {
      case 1:
        return GameState.User;
        break;
      case -1:
        return GameState.Random;
        break;
      case 0:
        return GameState.Draw;
        break;
      default:
        return GameState.Draw;
    }
  }

  render() {
    const {score} = this.props;
    const gameState = this.getGameState(score);

    return (
      <StyledWrapper>
        <StyledNum score={score[0]} gameState={gameState} user/>
        <StyledNum score={score[1]} gameState={gameState}/>
      </StyledWrapper>
    );
  }
}

export default Count;
