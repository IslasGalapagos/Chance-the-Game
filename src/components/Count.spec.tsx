import {shallow} from 'enzyme';
import * as testRenderer from 'react-test-renderer';
import * as React from 'react';
import Count, {StyledCountNum} from './Count';

describe('<Count/>', () => {
  it('renders correctly', () => {
    const treeAndStyles = testRenderer
      .create(<Count score={[0, 0]} />)
      .toJSON();
    expect(treeAndStyles).toMatchSnapshot();
  });

  it('renders with score from props', () => {
    const component = shallow(<Count score={[1, 2]} />);
    expect(
      component
        .find(StyledCountNum)
        .first()
        .render()
        .text()
    ).toEqual('1');
    expect(
      component
        .find(StyledCountNum)
        .last()
        .render()
        .text()
    ).toEqual('2');
  });

  it('overlap one element be another, depends on score', () => {
    const component = shallow(<Count score={[1, 1]} />);
    expect(
      component
        .find(StyledCountNum)
        .first()
        .prop('gameState')
    ).toEqual('draw');
    expect(
      component
        .find(StyledCountNum)
        .last()
        .prop('gameState')
    ).toEqual('draw');

    component.setProps({score: [2, 1]});
    expect(
      component
        .find(StyledCountNum)
        .first()
        .prop('gameState')
    ).toEqual('user_lead');
    expect(
      component
        .find(StyledCountNum)
        .last()
        .prop('gameState')
    ).toEqual('user_lead');

    const treeAndStyles = testRenderer.create(<Count score={[2, 1]} />);
    expect(treeAndStyles.toJSON()).toMatchSnapshot();

    component.setProps({score: [1, 2]});
    expect(
      component
        .find(StyledCountNum)
        .first()
        .prop('gameState')
    ).toEqual('random_lead');
    expect(
      component
        .find(StyledCountNum)
        .last()
        .prop('gameState')
    ).toEqual('random_lead');

    treeAndStyles.update(<Count score={[1, 2]} />);
    expect(treeAndStyles.toJSON()).toMatchSnapshot();
  });
});
