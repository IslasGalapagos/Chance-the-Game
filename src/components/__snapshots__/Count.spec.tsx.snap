// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Count/> overlap one element be another, depends on score 1`] = `
.emotion-0 {
  font-family: Arial;
  font-size: 36px;
  text-align: center;
  position: relative;
  display: inline-block;
  width: 64px;
  box-sizing: border-box;
  border-radius: 50%;
}

.emotion-4 {
  position: absolute;
  top: 35px;
  left: 35px;
}

.emotion-1 {
  line-height: 60px;
  left: 5px;
  z-index: 1;
  background-color: #FCFFDA;
  color: #49404E;
  border: 2px solid #49404E;
}

.emotion-3 {
  line-height: 64px;
  left: -5px;
  z-index: 0;
  background-color: #49404E;
  color: #FCFFDA;
}

<div
  className="emotion-4"
>
  <div
    className="emotion-0 emotion-1"
  >
    2
  </div>
  <div
    className="emotion-0 emotion-3"
  >
    1
  </div>
</div>
`;

exports[`<Count/> overlap one element be another, depends on score 2`] = `
.emotion-0 {
  font-family: Arial;
  font-size: 36px;
  text-align: center;
  position: relative;
  display: inline-block;
  width: 64px;
  box-sizing: border-box;
  border-radius: 50%;
}

.emotion-4 {
  position: absolute;
  top: 35px;
  left: 35px;
}

.emotion-1 {
  line-height: 60px;
  left: 5px;
  z-index: 0;
  background-color: #FCFFDA;
  color: #49404E;
  border: 2px solid #49404E;
}

.emotion-3 {
  line-height: 64px;
  left: -5px;
  z-index: 1;
  background-color: #49404E;
  color: #FCFFDA;
}

<div
  className="emotion-4"
>
  <div
    className="emotion-0 emotion-1"
  >
    1
  </div>
  <div
    className="emotion-0 emotion-3"
  >
    2
  </div>
</div>
`;

exports[`<Count/> renders correctly 1`] = `
.emotion-0 {
  font-family: Arial;
  font-size: 36px;
  text-align: center;
  position: relative;
  display: inline-block;
  width: 64px;
  box-sizing: border-box;
  border-radius: 50%;
}

.emotion-4 {
  position: absolute;
  top: 35px;
  left: 35px;
}

.emotion-1 {
  line-height: 60px;
  left: 1px;
  background-color: #FCFFDA;
  color: #49404E;
  border: 2px solid #49404E;
}

.emotion-3 {
  line-height: 64px;
  left: -1px;
  background-color: #49404E;
  color: #FCFFDA;
}

<div
  className="emotion-4"
>
  <div
    className="emotion-0 emotion-1"
  >
    0
  </div>
  <div
    className="emotion-0 emotion-3"
  >
    0
  </div>
</div>
`;
